### DISCUSSION TITLE
Question: Define extra keys for different layers

### DISCUSSION THREAD

**Discussion by mind33** [urob/zmk-config#31] [Q&A] [answered]
Hello,
I am building a sofle keyboard. How do I define the extra keys on the keyboard for different layers?

**Comment by urob** [comment]
Unfortunately, the current implementation isn't flexible enough for that. While it would certainly be possible to define layer-specific extra keys, it seems to me that at that point the additional complexity outweighs the gains of keeping all different keyboards in one repo and one may be better off splitting them up.

**Comment by mind33** [comment, marked as answer]
Manage to do it by renaming the variables for each layers. e.g. I rename X_LT to X_LT0, X_LT1, X_LT2 and so for the different layers. It's not very elegant but it does the job.
 I have encountered issue when adding the sensor bindings. Initially I did not add the bindings for the sensor and the rotary encoders button (when pressed) worked. However once I added the bindings for the encoders (which work fine), nothing happen when I press the rotary encoder buttons.

**Comment by urob** [comment]
Not sure about the encoder bindings. In case you use the `ZMK_LAYER` macro to set them up, the 3rd argument expands to `sensor-bindings` under the hood:

```cpp
#define ZMK_LAYER_3_ARGS(name, layout, sensors) \
    / { \
        keymap { \
            compatible = "zmk,keymap"; \
            layer_ ## name { \
                label = ZMK_HELPER_STRINGIFY(name); \
                bindings = <layout>; \
                sensor-bindings = <sensors>; \
            }; \
        }; \
    };
```

I don't use sensors myself but from what I understand this is the correct way to set them up. I'd check with the ZMK discord server to see if anyone knows something about the interaction between sensor-bindings and rotary buttons.

**Comment by auipga** [comment]
For those coming here like me:
The code above is not needed anymore. The feature is already there. Since the first implementation it has been [renamed](https://github.com/urob/zmk-helpers/commit/cce3bb12748c3f8b7317cc10f2edb15c9e017c28#diff-61f23ff8e2b460e7704ba04e4f23ad7134c0efd7392543c313f46086a3530dedR55) to `ZMK_LAYER_3` (without `_ARGS`)

**Reply by urob** [reply]
Note that you don't need to use `ZMK_LAYER_3` for this. This is just an under the hood implementation. You can specify sensors with the optional third argument using the regular `ZMK_LAYER`. See [its documentation](https://github.com/urob/zmk-helpers/blob/main/docs/core_helpers.md#zmk_layer) for details.
